dashboard/app/models/user_level.rb

```diff
@@ -130,28 +130,29 @@ def before_unsubmit
     end
   end
 
+  # `locked` is a virtual attribute because it relies on `unlocked_at` to
+  # automatically return `true` after `AUTOLOCK_PERIOD`
+  def locked
+    unlocked_at.nil? || unlocked_at < AUTOLOCK_PERIOD.ago
+  end
+
   def locked=(val)
-    new_time = if val
-                 nil
-               else
-                 Time.now
-               end
-    self.unlocked_at = new_time
-  end
-
-  def locked?
-    if unlocked_at.nil?
-      return true
-    else
-      return unlocked_at < AUTOLOCK_PERIOD.ago
-    end
+    self.unlocked_at = val ? nil : Time.now
   end
 
+  # this is the "locked" value we return to the client. if the lesson isn't
+  # lockable or we are an authorized teacher, we always return `false`.
   def show_as_locked?(stage)
     return false unless stage.lockable?
     return false if user.authorized_teacher?
-    return false if readonly_answers && !locked?
-    return locked? || submitted?
+    locked
+  end
+
+  # `readonly` and `locked` are mutually exclusive on the client, so we use
+  # this helper to override the value of `readonly_answers` when we're supposed
+  # to show as locked.
+  def show_as_readonly?(stage)
```

Comment:
i added this helper to parallel `show_as_locked`, to make explicit the fact that we want to override the `readonly_answers` value in the db if the ul is locked.

Reply:
This is nice